### PR TITLE
Merge webppl-nn

### LINKED ISSUE #710: Neural Net Helpers
(I'm splitting this out from #663.)

Idea: by leveraging our parameter/guide infrastructure we can use function composition to build simple neural networks in webppl. This would look something like this:

```js
// make a net:
var net = compose(sigmoid, linear(10, 'name'));
// use it:
net(ones([5,1]));
// slightly fuller example:
var predictNet = stack([linear(10, 'net1'), tanh, linear(2, 'net2'), sigmoid]);
var guideParams = {ps: predict(x)};
// etc.
```

Under the hood, the parameter infrastructure takes care of doing the right thing with parameters, and we determine whether we need to put a prior over parameters by checking whether we're in a guide thunk. (By building on #652.) ~~There's a work-in-progress implementation of this [here](https://github.com/probmods/webppl/compare/dev...null-a:wppl-nets).~~

For the user, the main advantage of this over using adnn nets is that there's no need to use `nneval` or call `setTraining` on the net. Also, with this approach, it's possible to mimic the common pattern of `cache`ing a fragment of network by simply reusing network names where sharing is required.

One question I'd like to answer (in part to help convince myself that this is flexible enough to handle future requirements) is how we best handle weight decay with this scheme. You can imagine specifying weight decay with something like `linear({..., weightDecay: 'l2'})`, and for nets that are part of the model we would implement this by switching from using `ImproperUniform` to `TensorGaussian` as the prior. But for nets that are part of the guide we don't have this option, so we need to do something else. (@dritchie and I have [discussed](https://gist.github.com/null-a/a42a855efbce27da1f8c4f4248ddd790) this a little already.)

### PR DESCRIPTION
**Pull request by null-a**
This merges [`webppl-nn`](https://github.com/null-a/webppl-nn) into core. I've added a few basic tests and tried to improve the documentation along the way.

Closes #710, #836.

@jpchen: Would merging this now cause problems with anything you're working on?

### CODE REVIEW

**Comment by jpchen** [comment]
Hm it would, but only slightly. I have a working webppl branch and a separate webppl-nn and moved some function calls like `relu()` and `lrelu()` to AD so that `nn` can be agnostic to Tensor type. Temporarily, people can just use that branch and `require` webppl-nn in the old way.
To not block anyone, this is good for merging by me, and I can just change the necessary functions in util.js after.

Side note: did you take out leaky relu? And for relu, is `nn.relu.eval(x)` any different than the `relu` [operation](https://github.com/null-a/webppl-nn/blob/master/src/nn.js#L39) in webppl-nn?

**Comment by null-a** [comment]
> did you take out leaky relu?

Yes. (l)relu seemed like something that would be better in adnn (or its successor) than webppl. I think we should add the leaky version to adnn (and add the trivial wrapper to webppl) if we want it.

> And for relu, is nn.relu.eval(x) any different than the relu operation in webppl-nn?

When writing webppl-nn I'd missed that adnn already had relu. It's only exposed as a "network" rather than an AD op, hence the `eval` call, but they're the same function.
